---
title: "Examining eQTLs in Hartwig cohort + post-hoc power analysis"
author: "Dr Jacob Househam"
date: "24/06/2021"
output: html_document
---

## Note that any Hartwig related data can only be shared with researchers who 
## have permission to access the Hartwig cohort data - please contact the authors for more information

```{r library, message=FALSE, warning=FALSE, include=FALSE}
library(GenomicRanges);library(data.table);library(dplyr);library(rtracklayer);library(VariantAnnotation);library(matrixStats);library(pwr);library(wesanderson)
options(datatable.fread.datatable=FALSE)
```

```{r function}
drop_multiallelic = function(d) {
  wh = elementNROWS(alt(d)) == 1
  if (any(!wh)) {
    warning(sprintf("Droping %d multiallelic sites!\n", sum(!wh)))
  }
  d[wh,]
}
```

## Convert eQTL data to hg19
```{r hg19_convert}
# Load eQTL data matrices
listmat <- readRDS('intermediates/eqtl_data_matrices.rds')
qtlids <- row.names(listmat$Mut)

# Track the mutation types (i.e. NS, Enh or Both)
allmut2gene <- listmat$Mut2Gene
qtltypes <- unlist(sapply(qtlids,function(x) { paste(unique(allmut2gene[which(allmut2gene$Locus==x),'Type']),collapse=';') } ))

# Convert to hg19
iddf <- data.frame(chr=gsub('^(chr\\S+):\\S+','\\1',qtlids),
                   start=gsub('^chr\\S+:(\\d+)_\\S+','\\1',qtlids),
                   end=gsub('^chr\\S+:(\\d+)_\\S+','\\1',qtlids))
idrange <- makeGRangesFromDataFrame(iddf)
chain <- import.chain("input_files/hg38Tohg19.over.chain")
hg19ids <- liftOver(idrange, chain)

# Remove loci for which equivalent 1:1 matching coordinates in hg19 could not be found
numfound <- unlist(lapply(hg19ids,length))
qtlids <- qtlids[which(numfound==1)];qtltypes <- qtltypes[which(numfound==1)]
hg19ids <- unlist(hg19ids[which(numfound==1)])

# Save hg19 coordinates of eQTL loci
hg19loci <- paste0(hg19ids@seqnames,':',hg19ids@ranges,'_',gsub('^chr\\S+:\\d+_(\\S+)','\\1',qtlids))
lifted <- data.frame(hg38=qtlids,hg19=hg19loci,type=qtltypes)
saveRDS(lifted,file='intermediates/lifted_ids.rds')
write.table(hg19loci,file='intermediates/hg19loci.txt',row.names=F,col.names=F,sep='\n',quote=F)
```

## Load mutations for Hartwig CRC mets and record overlap with eQTLs
```{r get_mutated_hartwig}
# Read in metadata and select for CRC metastases
meta <- fread('input/metadata_withstats.tsv',data.table=F)
crcmeta <- meta[which(meta$primaryTumorLocation=='Colon/Rectum'),];row.names(crcmeta) <- c(1:nrow(crcmeta))

# Load one vcf at a time, recording mutations present in eQTl analysis
mutmat <- matrix(0L,nrow=length(hg19loci),ncol=nrow(crcmeta))
row.names(mutmat) <- hg19loci;colnames(mutmat) <- crcmeta$sampleId
for(i in c(1:length(hwsamples))) {
  cursam <- crcmeta[i,]
  file <- paste0('<path_to_hartwig_somatic_data>',cursam$setName,'/',cursam$sampleId,'.purple.somatic.vcf.gz')
  suppressMessages(anno_vcf <- readVcf(file) %>% drop_multiallelic)
  
  # Add a 1 to the matrix to indicate presence of mutation in sample
  loci <- paste0('chr',row.names(anno_vcf))
  if(length(which(hg19loci %in% loci))>0) {
    mutmat[hg19loci %in% loci,cursam$sampleId] <- 1
  }
  
  # Progress log
  if((i %% 25)==0) {
    print(paste0('Loading and assessing mutations from sample ',cursam$sampleId,': ',i,'/',nrow(crcmeta),' (',signif(i/nrow(crcmeta)*100,2),'%)'))
  }
}

# Save completed mutation matrix
saveRDS(mutmat,file='intermediates/hartwig_qtl_mutmat.rds')
```

## Prepare Hartwig CRC met mutation and expression data
```{r prepare_to_analyse_HW}
# Load in eQTL analysis results and filter for loci that were convertable to hg19
mutidDF <- readRDS('results/eqtl_analysis_results.rds')
mutidDF <- mutidDF[which(mutidDF$Locus %in% lifted$hg38),];row.names(mutidDF) <- c(1:nrow(mutidDF))
row.names(mutmat) <- lifted$hg38
mutmat <- mutmat[mutidDF$Locus,]

# Get the number of mutated samples in HW for each eQTL locus analysed
mutidDF$HWrec <- rowSums(mutmat)

# Load in HW gene expression data
geneexp <- fread('<path_to_compiled_HW_adjTPM_expression_table>')
row.names(geneexp) <- geneexp$GeneID;geneexp <- geneexp[,c(2:ncol(geneexp))]
hwrnasam <- colnames(geneexp)

# Convert to log(TPM+1) and then into z-score
for(i in c(1:ncol(geneexp))) { geneexp[,i] <- log(geneexp[,i]+1) }
allexp <- rowMeans(geneexp);allsd <- rowSds(as.matrix(geneexp))
for(i in c(1:ncol(geneexp))) {
  geneexp[,i] <- (geneexp[,i]-allexp)/allsd
}

# Get the number of matched DNA-RNA HW samples mutated for each locus
mutrnamat <- mutmat[,hwrnasam]
mutidDF$HW_RNArec <- rowSums(mutrnamat)

# Order by decreasing samples mutated
mutidDF <- mutidDF[order(mutidDF$HW_RNArec,decreasing=T),];row.names(mutidDF) <- c(1:nrow(mutidDF))
```

## Analyse eQTLs via Hartwig CRC mets mutation and expression data
```{r analyse_HW_for_eQTLs}
# Filter for significant eQTLs and those which have gene expression data in HW
mutmut <- mutidDF[which(mutidDF$ModeladjP<0.01 & mutidDF$MutadjP<0.05 & mutidDF$Ensembl %in% row.names(geneexp)),];row.names(mutmut) <- c(1:nrow(mutmut))

# Analyse eQTL in HW (if >=2 mutated matched DNA-RNA samples in HW)
signcorrect <- c();mut_pvals <- c()
for(i in c(1:nrow(mutmut))) {
  curmut <- mutmut[i,];mut <- curmut$Locus
  curgen <- curmut$Ensembl
  
  if(curmut$HW_RNArec>=2) {
     geneData <- tmpData;geneData$Expression <- as.numeric(geneexp[curgen,])
    geneData[mutrnamat[mut,]==1,'Mut'] <- 'Mut';geneData$Mut <- factor(geneData$Mut,levels=c('WT','Mut'))
   
    res.lm <- lm(Expression~Mut+Purity,data=geneData);scope.lm <- lm(Expression~Purity,data=geneData)
    aov.lm <- anova(scope.lm,res.lm,test='LRT')
    pval <- ifelse(aov.lm$`Sum of Sq`[2]>0,aov.lm$`Pr(>Chi)`[2],1)
    signcorrect <- c(signcorrect,as.character(ifelse(sign(res.lm$coefficients[2])==sign(curmut$Mutes),'Correct','Incorrect')))
    mut_pvals <- c(mut_pvals,pval)
  } else {
    mut_pvals <- c(mut_pvals,1)
    signcorrect <- c(signcorrect,NA)
  }
}
mutmut$hw_pval <- mut_pvals
mutmut$SignCorrect <- signcorrect
```

## Plot significant HW eQTLs
### Figure S7 A-I
```{r plot_sig_HW_eQTLs}
# Filter for eQTLs which validate in HW
sigmut <- mutmut[which(mutmut$hw_pval<0.05 & mutmut$SignCorrect=='Correct' & mutmut$HW_RNArec>=2),]

# Plot Figure S7 A-I
pdf('figures/figS7AI.hw_mut_qtls_sigexp.pdf')
layout(matrix(c(1:4),nrow=2,byrow=T))
tmpData <- data.frame(Expression=rep(0,length(hwrnasam)),Mut='WT')
par(mar=c(4,3.5,2.5,1),font=2,font.axis=2,font.lab=2);options(scipen=-2)
for(i in c(1:nrow(sigmut))) {
  curmut <- sigmut[i,];mut <- curmut$Locus
  curgen <- curmut$Ensembl;curpval <- curmut$hw_pval
  
  geneData <- tmpData;geneData$Expression <- as.numeric(geneexp[curgen,])
  geneData[mutrnamat[mut,]==1,'Mut'] <- 'Mut';geneData$Mut <- factor(geneData$Mut,levels=c('WT','Mut'))
  
  maxexp <- ceiling(max(geneData$Expression));minexp <- floor(min(geneData$Expression))
  boxplot(Expression~Mut,data=geneData,
          ylim=c(minexp,maxexp),las=1,cex.axis=1.125,frame=F,outline=F,xlab='',ylab='',axes=F,boxcol=c('gray85',datacol[3]),boxlwd=2,col='white',medcol=c('gray70',datacol[3]),medlwd=3,staplecol=c('gray70',datacol[3]),staplelwd=2,outcol=c('gray70',datacol[3]),outpch=20)
  stripchart(Expression~Mut,data=geneData,vertical=T,method="jitter",add=TRUE,pch=20,col=scales::alpha(c('gray70',datacol[3]),0.5),cex=1)
  axis(side=1,at=c(1,2),labels=c('WT',paste0('Mut - ',curmut$Type)),cex.axis=1.5);axis(side=2,cex.axis=1.25,las=1,line=0)
  mtext(side=3,line=c(1,0),cex=c(1,0.75),font=2,text=c(paste0(curmut$Gene,' (',mut,')'),paste0('eQTL effect size = ',signif(curmut$Mutes,3),' Clonality = ',curmut$ClonalityRNA)))
  legend('topright',legend=paste0('p-value: ',signif(curpval,3)),bty='n',cex=0.65)
  mtext(side=2,line=2,text='Expression (z-score)')
  mtext(side=1,at=c(1,2),line=2,text=paste0('n=',as.numeric(table(geneData$Mut))))
}
dev.off()
```

## Run and plot post-hoc power analysis for HW cohort
### Figure S7J
```{r HW_eqtl_power}
# Select only loci which have at least 2 matched DNA-RNA HW samples mutated
mutmut <- mutmut[which(mutmut$HW_RNArec>=2),]
# Calculate power to detect each mutation change
sigthresh <- 0.05;samneeded <- c();options(scipen=100)
for(i in c(1:nrow(mutmut))) {
  curmut <- mutmut[i,]
  
  # If effect size is too large, power test returns NA
  es <- ifelse(abs(curmut$Mutes)>3.04,3.04,abs(curmut$Mutes))
  
  # Get number of samples needed based on number of WT samples
  # eQTL effect size, 0.05 sig level and 0.99 power
  respwr <- pwr.t2n.test(n2=length(hwrnasam)-curmut$HW_RNArec,d=es,sig.level=sigthresh,power=0.99)
  samneeded <- c(samneeded,respwr$n1*1.15)
}
mutmut$SamNeeded <- samneeded

# Plot Figure S7J
pdf('figures/figS7J.HW_eqtl_power_analysis.pdf')
par(mar=c(4,4,1,1),font=2,font.axis=2,font.lab=2,cex.axis=1.25)
plot(mutmut$SamNeeded[which(mutmut$hw_pval>=sigthresh | mutmut$SignCorrect!='Correct')],mutmut$HW_RNArec[which(mutmut$hw_pval>=sigthresh | mutmut$SignCorrect!='Correct')],axes=F,pch=16,xlab='',ylab='',
     col=scales::alpha('dimgray',0.5),ylim=c(0,15),xlim=c(0,20),cex=2)
points(mutmut$SamNeeded[which(mutmut$hw_pval<sigthresh & mutmut$SignCorrect=='Correct')],mutmut$HW_RNArec[which(mutmut$hw_pval<sigthresh & mutmut$SignCorrect=='Correct')],pch=18,cex=3,col=scales::alpha('firebrick3',0.8))
abline(a=0,b=1,untf=T,lty=2,col='black',lwd=3)
mtext(paste0('Samples needed (power=0.99, sig.level=',sigthresh,')'), side=1,line=2.5,cex=1.2)
mtext(paste0('Mutated HW DNA-RNA samples (total = ',length(hwrnasam),')'), side=2,line=2.5,cex=1.2)
axis(side=1);axis(side=2,las=2)
legend('topleft',bty='n',cex=1.2,legend=c('Non-significant in HW','Significant in HW'),pch=c(16,18),col=c(scales::alpha('dimgray',0.5),scales::alpha('firebrick3',0.8)))
dev.off()

```
