---
title: "Assessment of CMS and CRIS classification heterogeneity"
author: "Dr Jacob Househam"
date: "25/06/2021"
output: html_document
---

```{r library, message=FALSE, warning=FALSE, include=FALSE}
library(pheatmap);library(wesanderson);library(CMScaller);library(vegan)
getstars <- function(val) { if(val<0.001) {star<-'***'} else if(val<0.01) {star <-'**'} else if(val<0.05) { star<-'*' } else {star<-''} ;return(star)}
```

## Load colour palettes
```{r setup}
# CMS and CRIS colours obtained from original manuscripts
colcris <- c(A='#F7A143',B='#EE1F25',C='#3753A4',D='#6BBC42',E='#71C69F',Unk='#D4D4D4')
colcms <- c(CMS1='#E4E515',CMS2='#8789C2',CMS3='#C191C3',CMS4='#33B577',Unk='#D4D4D4')
ngroupcol= c(wes_palette("Darjeeling2",5)[2],wes_palette('Darjeeling1',5)[2],wes_palette('FantasticFox1',5)[c(5,2)])
names(ngroupcol) <- as.numeric(c(1:4))
```

## Load sample and gene expression data
```{r load_data}
# Load sample list
passsam <- read.table('input_files/ListRNAPass.EPICC.txt')[,1]
tsam <- passsam[-grep('^C\\d+_E',passsam)]
allpats <- gsub('^(C\\d+)_\\S+','\\1',tsam);patients <- unique(allpats);numsam <- table(allpats)
tsam <- tsam[which(allpats %in% names(which(numsam>=5)))]
allpats <- gsub('^(C\\d+)_\\S+','\\1',tsam);patients <- unique(allpats);numsam <- table(allpats)
msipat <- c('C516','C518','C536','C548','C552')
normsam <- passsam[grep('^C\\d+_E',passsam)];allnormpats <- gsub('^(C\\d+)_\\S+','\\1',normsam)
normsam <- normsam[which(allnormpats %in% allpats)]

# Load gene expression raw counts
counts <- readRDS('intermediates/All_EPICC_counts.rds')
row.names(counts) <- counts$GeneID

# Input pathway clustering data
genedf <- readRDS('results/gene_clustering_and_id_conversion.rds')
```

## Determine CMS and CRIS classifications
```{r get_cms_and_cris}
# Run CMScaller to get CMS and CRIS classes
forcris <- as.matrix(counts[,tsam])
cms <- CMScaller(forcris,templates = templates.CMS,rowNames='ensg',RNAseq=TRUE,nPerm=1000,FDR=0.05,doPlot = T,verbose = T)
cris <- CMScaller(forcris,templates = templates.CRIS,rowNames='ensg',RNAseq=TRUE,nPerm=1000,FDR=0.05,doPlot = T,verbose = T)
saveRDS(cms,'results/cms_raw_results.rds');saveRDS(cris,'results/cris_raw_results.rds')

# Concatenate calls
callingRes <- data.frame(Samples=row.names(cms),CMS='',CRIS='')
for(i in c(1:nrow(callingRes))) {
  curcms <- cms[which(row.names(cms)==callingRes[i,'Samples']),]
  callingRes[i,'CMS'] <- gsub('d.(CMS\\d)','\\1',names(which.min(curcms[,grep('CMS',colnames(curcms))])))
  curcris <- cris[which(row.names(cris)==callingRes[i,'Samples']),]
  callingRes[i,'CRIS'] <- gsub('d.(CRIS\\S)','\\1',names(which.min(curcris[,grep('CRIS',colnames(curcris))])))
}

# Organise results by patient 
patRes <- data.frame(CMS1=rep(0,length(patients)),CMS2=0L,CMS3=0L,CMS4=0L,CRISA=0L,CRISB=0L,CRISC=0L,CRISD=0L,CRISE=0L,NumSam=as.numeric(table(allpats)),row.names=patients)
for(i in c(1:length(patients))) {
  pat <- patients[i]
  patsam <- tsam[grep(pat,tsam)]
  for(cat in paste0('CMS',c(1:4))) {
    patRes[i,cat] <- length(which(callingRes[which(callingRes$Samples %in% patsam),'CMS']==cat))
  }
  for(cat2 in paste0('CRIS',c('A','B','C','D','E'))) {
    patRes[i,cat2] <- length(which(callingRes[which(callingRes$Samples %in% patsam),'CRIS']==cat2))
  }
}
```

## Plot results of CMS and CRIS classification
## Figure S2 A-C
```{r plot_cms_and_cris}
# Plot all CMS and CRIS results - Figure S1 A&B
pdf('figures/figS2AB.cms_and_cris_results.pdf')
par(mar=c(4.5,4.5,3,1),font=2,font.axis=2,cex.axis=1.35)
cmsplot <- t(patRes[,grep('CMS',colnames(patRes))])
barplot(cmsplot,col=colcms,border=NA,ylim=c(0,40),las=2)
mtext(side=2,text='Number of samples',line=2.75,font=2,xpd=T,cex=1.35)
text(x=10,y=42,labels='CMS',font=2,xpd=T,cex=1.4)
legend(x=0,y=41,bty='n',pt.cex=2,cex=1.2,fill=colcms[1:4],legend=c('CMS1','CMS2','CMS3','CMS4'),border=NA,xpd=T)

crisplot <- t(patRes[,grep('CRIS',colnames(patRes))])
barplot(crisplot,col=colcris,border=NA,ylim=c(0,40),las=2)
mtext(side=2,text='Number of samples',line=2.75,font=2,xpd=T,cex=1.35)
text(x=10,y=42,labels='CRIS',font=2,xpd=T,cex=1.4)
legend(x=0,y=41,bty='n',pt.cex=2,cex=1.1,fill=colcris[1:5],legend=c('CRIS-A','CRIS-B','CRIS-C','CRIS-D','CRIS-E'),border=NA,xpd=T)
dev.off()

# Get proportions per tumour for plotting
propRes <- patRes
for(i in c(1:nrow(patRes))) {
  for(j in c(1:ncol(patRes))) {
    propRes[i,j] <- patRes[i,j]/sum(patRes[i,])
  }
}

# Plot heatmap of CMS vs CRIS - Figure S1C
pdf('figures/figS2D.cmscris_heatmap.pdf')
pheatmap(t(patRes[,c(1:9)]),scale='column',fontsize=15,cellwidth=20,cellheight=40)
dev.off()
```

## Analyse and plot enrichment of gene groups vs classifications
## Figure S2D
```{r gene_groups_vs_classifications}
# Determine the number of genes used in CMS and CRIS classification
# which belong to each gene group
enrichCMS <- as.data.frame(matrix(0L,nrow=4,ncol=8));row.names(enrichCMS) <- c(1:4)
colnames(enrichCMS) <- c('NotCMSNotGroup','NotCMSGroup','CMSNotGroup','CMSGroup','OR','lowOR','highOR','pval')
enrichCRIS <- enrichCMS;colnames(enrichCRIS) <- c('NotCRISNotGroup','NotCRISGroup','CRISNotGroup','CRISGroup','OR','lowOR','highOR','pval')
for(i in c(1:4)) {
  test <- i
  gengroup <- genedf$entrezgene[which(genedf$Group==as.character(i))]
  gennongroup <- genedf$entrezgene[which(genedf$Group!=as.character(i))]
  testmat <- rbind(c(length(which(gennongroup %ni% unique(templates.CMS$probe))),length(which(gengroup %ni% unique(templates.CMS$probe)))),c(length(which(gennongroup %in% unique(templates.CMS$probe))),length(which(gengroup %in% unique(templates.CMS$probe)))))
  enrichCMS[i,'NotCMSNotGroup'] <- length(which(gennongroup %ni% unique(templates.CMS$probe)));enrichCMS[i,'NotCMSGroup'] <-length(which(gengroup %ni% unique(templates.CMS$probe)))
  enrichCMS[i,'CMSNotGroup'] <-length(which(gennongroup %in% unique(templates.CMS$probe)));enrichCMS[i,'CMSGroup'] <-length(which(gengroup %in% unique(templates.CMS$probe)))
  res <- fisher.test(testmat)
  enrichCMS[i,'OR'] <- res$estimate;enrichCMS[i,'lowOR'] <- res$conf.int[1];enrichCMS[i,'highOR'] <- res$conf.int[2]
  enrichCMS[i,'pval'] <- res$p.value
  
  testmat <- rbind(c(length(which(gennongroup %ni% unique(templates.CRIS$probe))),length(which(gengroup %ni% unique(templates.CRIS$probe)))),c(length(which(gennongroup %in% unique(templates.CRIS$probe))),length(which(gengroup %in% unique(templates.CRIS$probe)))))
  enrichCRIS[i,'NotCRISNotGroup'] <-length(which(gennongroup %ni% unique(templates.CRIS$probe)));enrichCRIS[i,'NotCRISGroup'] <-length(which(gengroup %ni% unique(templates.CRIS$probe)))
  enrichCRIS[i,'CRISNotGroup'] <-length(which(gennongroup %in% unique(templates.CRIS$probe)));enrichCRIS[i,'CRISGroup'] <-length(which(gengroup %in% unique(templates.CRIS$probe)))
  res <- fisher.test(testmat,conf.level = 0.95)
  enrichCRIS[i,'OR'] <- res$estimate;enrichCRIS[i,'lowOR'] <- res$conf.int[1];enrichCRIS[i,'highOR'] <- res$conf.int[2]
  enrichCRIS[i,'pval'] <- res$p.value
}
sigs <- c();for(p in p.adjust(c(enrichCMS$pval,enrichCRIS$pval),method='fdr')) { sigs <- c(sigs,getstars(p)) }

# Plot Figure S1D
pdf('figures/figS2D.groupenrichmentCMSCRIS.pdf')
par(mar=c(5,5.1,1,0.6),font.lab=2,cex.axis=1.5,font.axis=2,font=2)
xx <- barplot(rbind(enrichCMS$OR,enrichCRIS$OR),beside=T,axes=F,las=2,density=rep(c(NA,20),4),
              col=rep(ngroupcol,each=2),border=rep(ngroupcol,each=2),ylim=c(0,2.2));abline(h=1,lty=3,lwd=2)
axis(side=2,at=c(0,1,2),labels=c(0,1,2),lwd=2,las=2)
mtext(side=2,line=3,text='OR: CMS/CRIS genes in groups',cex=1.35)
axis(side=1,at=(xx[2,]-((xx[2,]-xx[1,])/2)),labels = c(1:4),lwd=2,line=1)
mtext(side=1,line=3.5,text='Gene groups',cex=1.35)
segments(x0=xx[1,],y0=enrichCMS$lowOR,y1=enrichCMS$highOR,lwd=4)
segments(x0=xx[2,],y0=enrichCRIS$lowOR,y1=enrichCRIS$highOR,lwd=4)
text(x=c(xx[1,],xx[2,]),y=0.1,labels=sigs,cex=1.5,font=2)
legend(x=0.5,y=2.1,legend=c('CMS','CRIS'),cex=2,fill='gray60',bty='n',density=c(NA,20),border=c('gray60'))
dev.off()
```
